docs/core/middleware.md

```diff
@@ -0,0 +1,155 @@
+
+Where we define the following arguments and return types:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/master/types/tx/middleware.go#L25-L60
+
+BaseApp holds a reference to a `tx.Handler`:
+
+```go
+type BaseApp  struct {
+    // other fields
+    txHandler tx.Handler
+}
+```
+
+
+## Implementing a Middleware
+
+To implement a middleware is done simply by a Go function that takes as arguments and some parameters and returns a tx. Middleware.
+
+For example, we can create a Generic middleware. 
+
+```go
+// myTxHandler is the tx.Handler of this middleware. Note that it holds a
+// reference to the next tx.Handler in the stack.
+type myTxHandler struct {
+    // next is the next tx.Handler in the middleware stack.
+    next tx.Handler
+    // some other fields that are relevant to the middleware can be added here
+}
+
+// NewMyMiddleware returns a middleware that does this and that.
+func NewMyMiddleware(arg1, arg2) tx.Middleware {
+    return func (txh tx.Handler) tx.Handler {
+        return myTxHandler{
+            next: txh,
+            // optionally, set arg1, arg2... if they are needed in the middleware
+        }
+    }
+}
+
+// Assert myTxHandler is a tx.Handler.
+var _ tx.Handler = myTxHandler{}
+```
+
+Next we need to implement the middleware [interface](#Type-Definition).
+
+```go
+func (h myTxHandler) CheckTx(ctx context.Context, req Request, checkReq RequestcheckTx) (Response, ResponseCheckTx, error) {
+    // CheckTx specific pre-processing logic
+
+    // run the next middleware
+    res, checkRes, err := txh.next.CheckTx(ctx, req, checkReq)
+
+    // CheckTx specific post-processing logic
+
+    return res, checkRes, err
+}
+
+func (h myTxHandler) DeliverTx(ctx context.Context, req Request) (Response, error) {
+    // DeliverTx specific pre-processing logic
+
+    // run the next middleware
+    res, err := txh.next.DeliverTx(ctx, tx, req)
+
+    // DeliverTx specific post-processing logic
+
+    return res, err
+}
+
+func (h myTxHandler) SimulateTx(ctx context.Context, req Request) (Response, error) {
+    // SimulateTx specific pre-processing logic
+
+    // run the next middleware
+    res, err := txh.next.SimulateTx(ctx, tx, req)
+
+    // SimulateTx specific post-processing logic
+
+    return res, err
+}
+```
+
+### Composing Middlewares
+
+While BaseApp holds a reference to a `tx.Handler`, this `tx.Handler` itself is defined using a middleware stack. The Cosmos SDK exposes a base (i.e. innermost) `tx.Handler` called `RunMsgsTxHandler`, which executes messages.
+
+Then, the app developer can compose multiple middlewares on top on the base `tx.Handler`. Each middleware can run pre-and-post-processing logic around its next middleware, as described in the section above. Conceptually, as an example, given the middlewares `A`, `B`, and `C` and the base `tx.Handler` `H` the stack looks like:
+
+![Composing](baseapp_transaction-middleware.png)
```

Comment:
@amaurym Also I added this image but I don't know if is correct for the cosmos SDK branding, will you like me to change color, shape or remove it altogether

Reply:
We currently don't have any other "owning" the documentation, so there's no uniform branding here. I really like the image! With what program did you do this? If it's some kind of UML diagram, maybe it's worth adding the source too ? (like we did in docs/uml).

Reply:
It is lucidshart, I added the reference let me know what you think

Reply:
i added the reference let me know what do you think

Reply:
I see. Yeah let's actually remove the reference, I clicked your link, but it seems that I need to sign up first.

Let's jsut keep the png in this folder.